File: src/jp/mcbe/fuyutsuki/Texter/data/FloatingTextData.php
<?php

/**
 * // English
 *
 * Texter, the display FloatingTextPerticle plugin for PocketMine-MP
 * Copyright (c) 2019-2021 yuko fuyutsuki < https://github.com/fuyutsuki >
 *
 * This software is distributed under "NCSA license".
 * You should have received a copy of the NCSA license
 * along with this program.  If not, see
 * < https://opensource.org/licenses/NCSA >.
 *
 * ---------------------------------------------------------------------
 * // 日本語
 *
 * TexterはPocketMine-MP向けのFloatingTextPerticleを表示するプラグインです
 * Copyright (c) 2019-2021 yuko fuyutsuki < https://github.com/fuyutsuki >
 *
 * このソフトウェアは"NCSAライセンス"下で配布されています。
 * あなたはこのプログラムと共にNCSAライセンスのコピーを受け取ったはずです。
 * 受け取っていない場合、下記のURLからご覧ください。
 * < https://opensource.org/licenses/NCSA >
 */

declare(strict_types=1);

namespace jp\mcbe\fuyutsuki\Texter\data;

use jp\mcbe\fuyutsuki\Texter\task\PrepareTextsTask;
use jp\mcbe\fuyutsuki\Texter\text\FloatingTextCluster;
use jp\mcbe\fuyutsuki\Texter\util\StringArrayMultiton;
use pocketmine\plugin\PluginBase;
use pocketmine\utils\Config;

class FloatingTextData extends Config {

	use StringArrayMultiton {
		StringArrayMultiton::__construct as stringArrayMultitonConstruct;
	}

	public const FLOATING_TEXT_DIRECTORY = "floating_text" . DIRECTORY_SEPARATOR;

	private string $folderName;
	/** @var FloatingTextCluster[] */
	private array $floatingTexts = [];

	public function __construct(PluginBase $plugin, string $worldFolderName) {
		$floatingTextDir = $plugin->getDataFolder() . self::FLOATING_TEXT_DIRECTORY;
		parent::__construct($floatingTextDir . "$worldFolderName.json", Config::JSON);
		$this->setJsonOptions(Data::JSON_OPTIONS);
		$this->stringArrayMultitonConstruct($worldFolderName);
		$this->folderName = $worldFolderName;
	}

	public function generateFloatingTexts(PluginBase $plugin) {
		$prepare = new PrepareTextsTask($plugin, $this);
		$plugin->getScheduler()->scheduleRepeatingTask($prepare, PrepareTextsTask::TICKING_PERIOD);
	}

	public function folderName(): string {
		return $this->folderName;
	}

	public function floatingTexts(): array {
		return $this->floatingTexts;
	}

	public function floatingText(string $name): ?FloatingTextCluster {
		return $this->floatingTexts[$name] ?? null;
	}

	public function store(FloatingTextCluster $text, bool $onlyCache = false) {
		$name = $text->name();
		$this->floatingTexts[$name] = $text;
		if (!$onlyCache) {
			$this->set($name, $text);
		}
	}

	public function removeFloatingText(string $name, bool $onlyCache = false) {
		unset($this->floatingTexts[$name]);
		if (!$onlyCache) {
			$this->remove($name);
		}
	}

	public function existsFloatingText(string $name, bool $onlyCache = false): bool {
		$result = isset($this->floatingTexts[$name]);
		if (!$onlyCache) {
			$result = $result && $this->exists($name, true);
		}
		return $result;
	}

	public function notExistsFloatingText(string $name, bool $onlyCache = false): bool {
		return !$this->existsFloatingText($name, $onlyCache);
	}

}

File: src/main/java/jp/mcbe/fuyutsuki/Texter/data/FloatingTextData.java
package jp.mcbe.fuyutsuki.Texter.data;

import jp.mcbe.fuyutsuki.Texter.task.PrepareTextsTask;
import jp.mcbe.fuyutsuki.Texter.text.FloatingTextCluster;
import jp.mcbe.fuyutsuki.Texter.util.StringArrayMultitonRegistry;
import org.powernukkitx.plugin.PluginBase;
import org.powernukkitx.utils.Config;

import java.io.File;
import java.util.LinkedHashMap;
import java.util.Map;

public final class FloatingTextData extends Config {

	public static final String FLOATING_TEXT_DIRECTORY = "floating_text" + File.separator;

	private static final StringArrayMultitonRegistry<FloatingTextData> INSTANCES = new StringArrayMultitonRegistry<>();

	private final String folderName;

	private final Map<String, FloatingTextCluster> floatingTexts = new LinkedHashMap<>();

	public FloatingTextData(PluginBase plugin, String worldFolderName) {
		super(plugin.getDataFolder() + FLOATING_TEXT_DIRECTORY + worldFolderName + ".json", Config.JSON);

		this.folderName = worldFolderName;
		INSTANCES.register(worldFolderName, this);
	}

	public static FloatingTextData getInstance(String key) {
		return INSTANCES.get(key);
	}

	public static void removeInstance(String key) {
		INSTANCES.remove(key);
	}

	public void generateFloatingTexts(PluginBase plugin) {
		PrepareTextsTask prepare = new PrepareTextsTask(plugin, this);
		plugin.getServer().getScheduler().scheduleRepeatingTask(
			plugin, prepare, PrepareTextsTask.TICKING_PERIOD
		);
	}

	public String folderName() {
		return folderName;
	}

	public Map<String, FloatingTextCluster> floatingTexts() {
		return floatingTexts;
	}

	public FloatingTextCluster floatingText(String name) {
		return floatingTexts.get(name);
	}

	public void store(FloatingTextCluster text, boolean onlyCache) {
		String name = text.name();
		floatingTexts.put(name, text);
		if (!onlyCache) {
			this.set(name, text);
		}
	}

	public void removeFloatingText(String name, boolean onlyCache) {
		floatingTexts.remove(name);
		if (!onlyCache) {
			this.remove(name);
		}
	}

	public boolean existsFloatingText(String name, boolean onlyCache) {
		boolean result = floatingTexts.containsKey(name);
		if (!onlyCache) {
			result = result && this.exists(name, true);
		}
		return result;
	}

	public boolean notExistsFloatingText(String name, boolean onlyCache) {
		return !existsFloatingText(name, onlyCache);
	}
}